pkg/media/h264reader/h264reader.go

```diff
@@ -0,0 +1,272 @@
+// parser code written by https://github.com/chertov
+
+package h264reader
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"strconv"
+)
+
+// H264Reader reads data from stream and constructs h264 nal units
+type H264Reader struct {
+	stream io.Reader
+}
+
+// NewReader creates new H264Reader
+func NewReader(in io.Reader) (*H264Reader, error) {
+	if in == nil {
+		return nil, fmt.Errorf("stream is nil")
+	}
+
+	reader := &H264Reader{
+		stream: in,
+	}
+
+	return reader, nil
+}
+
+// NAL H.264 Network Abstraction Layer
+type NAL struct {
+	PictureOrderCount uint32
+
+	// NAL header
+	ForbiddenZeroBit bool
+	RefIdc           uint8
+	UnitType         nalUnitType
+
+	Data []byte // header byte + rbsp
+}
+
+// ReadFrames reads all data from stream and returns array of all parsed nal units
+func (reader *H264Reader) ReadFrames() []NAL {
```

Comment:
I was playing with this, and I think we need to re-think this :( I wanted to do `ffmpeg stdio -> pion` and reading all up front doesn't work. All of our other Reader APIs do a frame at a time.

I think we need to refactor to read 1 byte at a time, and then emit when we read a full frame.

* First run assert it starts with `0x00, 0x00, 0x01`
* Read until you hit `0x00, 0x00, 0x01` again, return the frame.

Then all the next reads just go until they hit `0x00, 0x00, 0x01` and we just keep some local state (already seen NAL header)

sorry if this is a pain, if you want we can push a V1 and then do a refactor if you are out of time on this :)

Reply:
Thanks for comments. I will refactor as you suggest.

Reply:
done